impl/maven-impl/src/main/java/org/apache/maven/impl/PathSelector.java

```diff
@@ -0,0 +1,475 @@
+            switch (includes.length) {
+                case 0:
+                    return Optional.empty();
+                case 1:
+                    return Optional.of(includes[0]);
+            }
+        }
+        return Optional.of(this);
+    }
+
+    /**
+     * Determines whether a path is selected.
+     * This is true if the given file matches an include pattern and no exclude pattern.
+     *
+     * @param path the pathname to test, must not be {@code null}
+     * @return {@code true} if the given path is selected, {@code false} otherwise
+     */
+    @Override
+    public boolean matches(Path path) {
+        path = baseDirectory.relativize(path);
+        return (includes.length == 0 || isMatched(path, includes))
+                && (excludes.length == 0 || !isMatched(path, excludes));
+    }
+
+    /**
+     * {@return whether the given file matches according one of the given matchers}.
+     */
+    private static boolean isMatched(Path path, PathMatcher[] matchers) {
+        for (PathMatcher matcher : matchers) {
+            if (matcher.matches(path)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether a directory could contain selected paths.
+     *
+     * @param directory the directory pathname to test, must not be {@code null}
+     * @return {@code true} if the given directory might contain selected paths, {@code false} if the
+     *         directory will definitively not contain selected paths
+     */
+    public boolean couldHoldSelected(Path directory) {
+        if (baseDirectory.equals(directory)) {
+            return true;
+        }
+        directory = baseDirectory.relativize(directory);
+        return (dirIncludes.length == 0 || isMatched(directory, dirIncludes))
+                && (dirExcludes.length == 0 || !isMatched(directory, dirExcludes));
+    }
+
+    /**
+     * Appends the elements of the given array in the given buffer.
+     * This is a helper method for {@link #toString()} implementations.
+     *
+     * @param buffer the buffer where to add the elements
+     * @param label label identifying the array of elements to add
+     * @param patterns the elements to append, or {@code null} if none
+     */
+    private static void append(StringBuilder buffer, String label, String[] patterns) {
+        buffer.append(label).append(": [");
+        if (patterns != null) {
+            for (int i = 0; i < patterns.length; i++) {
+                if (i != 0) {
+                    buffer.append(", ");
+                }
+                buffer.append(patterns[i]);
+            }
+        }
+        buffer.append(']');
+    }
+
+    /**
+     * {@return a string representation for logging purposes}.
+     * This string representation is reported logged when {@link Cleaner} is executed.
+     */
+    @Override
+    public String toString() {
+        var buffer = new StringBuilder();
```

Comment:
personally I dislike var. Not sure if there's a project setting on this

Reply:
I do not know if there is a Maven policy. Personally, I use it when it is redundant with what we can read on the same line with no possibility of different result (e.g., the result of `new`), but generally avoid it for the return value of a method call for example, as it is not obvious by reading the line.

Reply:
ideally we could leverage https://kotlinlang.org/docs/basic-syntax.html#variables or https://projectlombok.org/features/val to get rid of `boilerplate`; obtain beloved `immutability`.

https://discuss.kotlinlang.org/t/why-not-have-everything-be-val-by-default/25391

Reply:
> ideally we could leverage `val` https://kotlinlang.org/docs/basic-syntax.html#variables or supplementation https://projectlombok.org/features/val, to get rid of `boilerplate`; obtain beloved `immutability`.
> 
> https://discuss.kotlinlang.org/t/why-not-have-everything-be-val-by-default/25391

We don't use kotlin, nor Lombok, so no.

Reply:
Many languages use `val` and `var` as type it not first priority. less seems more to me.

- https://github.com/apache/maven/pull/2326
- 

<img width="755" alt="image" src="https://github.com/user-attachments/assets/201ac7a1-34ed-4156-a067-bd26eb11ae1b" />